File: sesion2-lab/spring-security/resource-server/src/main/java/com/example/demo/web/ApiController.java
package com.example.demo.web;

import java.util.List;
import java.util.Map;

import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.oauth2.jwt.Jwt;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/api")
public class ApiController {

    @GetMapping("/public/hello")
    public Map<String, Object> publicHello() {
        return Map.of(
            "message", "Hola! Este endpoint es publico, no requiere token.",
            "scope", "public");
    }

    @GetMapping("/me")
    public Map<String, Object> me(@AuthenticationPrincipal Jwt jwt, Authentication authentication) {
        List<String> authorities = authentication.getAuthorities().stream()
            .map(GrantedAuthority::getAuthority)
            .toList();
        return Map.of(
            "message", "Estas autenticado.",
            "user", jwt.getClaimAsString("preferred_username"),
            "email", String.valueOf(jwt.getClaimAsString("email")),
            "authorities", authorities);
    }

    @GetMapping("/admin/hello")
    public Map<String, Object> adminHello(@AuthenticationPrincipal Jwt jwt) {
        return Map.of(
            "message", "Bienvenido al area de administracion.",
            "user", jwt.getClaimAsString("preferred_username"),
            "scope", "admin");
    }
}

File: sesion2-lab/aspnet-identity-keycloak/identity-demo/Controllers/ApiController.cs
using System.Security.Claims;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace IdentityDemo.Controllers;

[ApiController]
[Route("api")]
public class ApiController : ControllerBase
{
    [HttpGet("public/hello")]
    [AllowAnonymous]
    public IActionResult PublicHello()
    {
        return Ok(new
        {
            message = "Hola! Este endpoint es publico, no requiere token.",
            scope = "public"
        });
    }

    [HttpGet("me")]
    [Authorize]
    public IActionResult Me()
    {
        var username = User.Identity?.Name
            ?? User.FindFirstValue("preferred_username");
        var email = User.FindFirstValue("email");
        var authorities = User.FindAll(ClaimTypes.Role).Select(c => "ROLE_" + c.Value).ToList();

        return Ok(new
        {
            message = "Estas autenticado con un token de Keycloak.",
            user = username,
            email,
            authorities
        });
    }

    [HttpGet("admin/hello")]
    [Authorize(Roles = "ADMIN")]
    public IActionResult AdminHello()
    {
        var username = User.Identity?.Name
            ?? User.FindFirstValue("preferred_username");

        return Ok(new
        {
            message = "Bienvenido al area de administracion.",
            user = username,
            scope = "admin"
        });
    }
}
